utxoscanner.go

```diff
@@ -0,0 +1,409 @@
+		if err != nil {
+			log.Errorf("utxo scan failed: %v", err)
+		}
+
+		select {
+		case <-s.quit:
+			return
+		default:
+		}
+	}
+}
+
+// dequeueAtHeight returns all GetUtxoRequests that have starting height of the
+// given height.
+func (s *UtxoScanner) dequeueAtHeight(height uint32) []*GetUtxoRequest {
+	s.cv.L.Lock()
+	defer s.cv.L.Unlock()
+
+	// Take any requests that are too old to go in this batch and keep them for
+	// the next batch.
+	for !s.pq.IsEmpty() && s.pq.Peek().StartHeight < height {
+		item := heap.Pop(&s.pq).(*GetUtxoRequest)
+		s.nextBatch = append(s.nextBatch, item)
+	}
+
+	var requests []*GetUtxoRequest
+	for !s.pq.IsEmpty() && s.pq.Peek().StartHeight == height {
+		item := heap.Pop(&s.pq).(*GetUtxoRequest)
+		requests = append(requests, item)
+	}
+
+	return requests
+}
+
+// scan runs a single batch, pulling in any requests that get added above the
+// batch's last processed height. If there was an error, then return the
+// outstanding requests.
+func (s *UtxoScanner) scanFromHeight(initHeight uint32) error {
+	// Before beginning the scan, grab the best block stamp we know of,
+	// which will serve as an initial estimate for the end height of the
+	// scan.
+	bestStamp, err := s.cfg.BestSnapshot()
+	if err != nil {
+		return err
+	}
+
+	var (
+		// startHeight and endHeight bound the range of the current
+		// scan. If more blocks are found while a scan is running,
+		// these values will be updated afterwards to scan for the new
+		// blocks.
+		startHeight = initHeight
+		endHeight   = uint32(bestStamp.Height)
+	)
+
+	reporter := newBatchSpendReporter()
+
+scanToEnd:
+	// Scan forward through the blockchain and look for any transactions that
+	// might spend the given UTXOs.
+	for height := startHeight; height <= endHeight; height++ {
+		// Before beginning to scan this height, check to see if the
+		// utxoscanner has been signaled to exit.
+		select {
+		case <-s.quit:
+			return reporter.FailRemaining(ErrShuttingDown)
+		default:
+		}
+
+		hash, err := s.cfg.GetBlockHash(int64(height))
+		if err != nil {
+			return reporter.FailRemaining(err)
+		}
+
+		// If there are any new requests that can safely be added to this batch,
+		// then try and fetch them.
+		newReqs := s.dequeueAtHeight(height)
+
+		// If an outpoint is created in this block, then fetch it regardless.
+		// Otherwise check to see if the filter matches any of our watched
+		// outpoints.
+		fetch := len(newReqs) > 0
+		if !fetch {
+			options := rescanOptions{
+				watchList: reporter.filterEntries,
+			}
+
+			match, err := s.cfg.BlockFilterMatches(&options, hash)
+			if err != nil {
+				return reporter.FailRemaining(err)
+			}
+
+			// If still no match is found, we have no reason to
+			// fetch this block, and can continue to next height.
+			if !match {
+				continue
+			}
+		}
+
+		// At this point, we've determined that we either (1) have new
+		// requests which we need the block to scan for originating
```

Comment:
I don't think it is obvious that `startHeight` is assumed to be the height at which the utxo is created.